model_hub/model_hub/huggingface/_trial.py

```diff
@@ -0,0 +1,256 @@
+        datasets = load_dataset(data_config.dataset_name, data_config.dataset_config_name)
+        if "validation" not in datasets.keys():
+            assert (
+                "validation_split_percentage" in data_config
+            ), "Validation split not provided by this huggingface dataset. Please specify validation_split_percentage in data_config for use to create validation set"
+            datasets["validation"] = load_dataset(
+                data_config.dataset_name,
+                data_config.dataset_config_name,
+                split=f"train[:{data_config.validation_split_percentage}%]",
+            )
+            datasets["train"] = load_dataset(
+                data_config.dataset_name,
+                data_config.dataset_config_name,
+                split=f"train[{data_config.validation_split_percentage}%:]",
+            )
+    else:
+        data_files = {}
+        if data_config.train_file is not None:
+            data_files["train"] = data_config.train_file
+        if data_config["validation_file"] is not None:
+            data_files["validation"] = data_config.validation_file
+        extension = data_config.train_file.split(".")[-1]
+        if extension == "txt":
+            extension = "text"
+        datasets = load_dataset(extension, data_files=data_files)
+    return datasets
+
+
+class BaseTransformerTrial(PyTorchTrial):
+    """
+    This is the base trial class for transformers with a default init and train_batch method.
+
+    You can subclass BaseTransformerTrial to customize a trial for your own usage by filing in
+    the expected methods for data loading and evaluation.
+
+    See examples/huggingface/token-classification/ner_trial.py for an example.
+    """
+
+    def __init__(self, context: PyTorchTrialContext) -> None:
+        self.context = context
+        self.hparams = AttrDict(context.get_hparams())
+        self.data_config = AttrDict(context.get_data_config())
+
+        # Parse hparams and data_config.
+        config_args, tokenizer_args, model_args = default_parse_config_tokenizer_model_args(
+            self.hparams
+        )
+        optimizer_args, scheduler_args = default_parse_optimizer_lr_scheduler_args(self.hparams)
+
+        self.config, self.tokenizer, self.model = build_using_auto(
+            config_args, tokenizer_args, self.hparams.model_mode, model_args
+        )
+        self.model = self.context.wrap_model(self.model)
+
+        self.optimizer = self.context.wrap_optimizer(
+            build_default_optimizer(self.model, optimizer_args)
+        )
+
+        if self.hparams.use_apex_amp:
+            self.model, self.optimizer = self.context.configure_apex_amp(
+                models=self.model,
+                optimizers=self.optimizer,
+            )
+
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
+            build_default_lr_scheduler(self.optimizer, scheduler_args),
+            LRScheduler.StepMode.STEP_EVERY_BATCH,
```

Comment:
Do we want to allow users to set this StepMode or does Huggingface not have that abstraction?

Reply:
our BaseTransformerTrial will basically replicate the same training loop as Trainer.  

For HF Trainer, all lr schedulers returned from transformer's get_scheduler method step every batch:
https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L1094